File: src/main/java/org/cubexmc/railway/util/MathUtil.java
package org.cubexmc.railway.util;

/**
 * Math utilities from BKCommonLib
 * Simplified version for Railway plugin use
 */
public class MathUtil {

    public static int floor(double value) {
        int i = (int) value;
        return value < i ? i - 1 : i;
    }

    public static int ceil(double value) {
        int i = (int) value;
        return value > i ? i + 1 : i;
    }

    public static double distance(double x1, double y1, double z1, double x2, double y2, double z2) {
        return Math.sqrt(distanceSquared(x1, y1, z1, x2, y2, z2));
    }

    public static double distanceSquared(double x1, double y1, double z1, double x2, double y2, double z2) {
        double dx = x2 - x1;
        double dy = y2 - y1;
        double dz = z2 - z1;
        return dx * dx + dy * dy + dz * dz;
    }

    public static double getNormalizationFactor(double x, double y, double z) {
        double len = Math.sqrt(x * x + y * y + z * z);
        return len < 1e-10 ? Double.POSITIVE_INFINITY : 1.0 / len;
    }

    public static double round(double value, int decimals) {
        double p = Math.pow(10, decimals);
        return Math.round(value * p) / p;
    }

    public static double clamp(double value, double min, double max) {
        return value < min ? min : (value > max ? max : value);
    }

    public static int clamp(int value, int min, int max) {
        return value < min ? min : (value > max ? max : value);
    }

    public static boolean isHeadingTo(double dx, double dz, org.bukkit.block.BlockFace direction) {
        return isHeadingTo(direction.getModX(), direction.getModZ(), dx, dz);
    }

    public static boolean isHeadingTo(double cx, double cz, double dx, double dz) {
        return cx * dx + cz * dz > 0.0;
    }

    public static double lerp(double a, double b, double t) {
        return a + (b - a) * t;
    }
}

File: src/main/java/org/cubexmc/metro/util/MathUtil.kt
package org.cubexmc.metro.util

import org.bukkit.block.BlockFace
import kotlin.math.roundToLong
import kotlin.math.sqrt

/**
 * Math utilities from BKCommonLib
 * Simplified version for Metro plugin use
 */
object MathUtil {

    private const val NORMALIZATION_EPSILON = 1e-10

    @JvmStatic
    fun floor(value: Double): Int {
        val i = value.toInt()
        return if (value < i) i - 1 else i
    }

    @JvmStatic
    fun ceil(value: Double): Int {
        val i = value.toInt()
        return if (value > i) i + 1 else i
    }

    @JvmStatic
    fun distance(x1: Double, y1: Double, z1: Double, x2: Double, y2: Double, z2: Double): Double =
        sqrt(distanceSquared(x1, y1, z1, x2, y2, z2))

    @JvmStatic
    fun distanceSquared(x1: Double, y1: Double, z1: Double, x2: Double, y2: Double, z2: Double): Double {
        val dx = x2 - x1
        val dy = y2 - y1
        val dz = z2 - z1
        return dx * dx + dy * dy + dz * dz
    }

    @JvmStatic
    fun getNormalizationFactor(x: Double, y: Double, z: Double): Double {
        val len = sqrt(x * x + y * y + z * z)
        return if (len < NORMALIZATION_EPSILON) Double.POSITIVE_INFINITY else 1.0 / len
    }

    @JvmStatic
    fun round(value: Double, decimals: Int): Double {
        val p = Math.pow(10.0, decimals.toDouble())
        return (value * p).roundToLong() / p
    }

    @JvmStatic
    fun clamp(value: Double, min: Double, max: Double): Double =
        if (value < min) min else if (value > max) max else value

    @JvmStatic
    fun clamp(value: Int, min: Int, max: Int): Int =
        if (value < min) min else if (value > max) max else value

    @JvmStatic
    fun isHeadingTo(dx: Double, dz: Double, direction: BlockFace): Boolean =
        isHeadingTo(direction.modX.toDouble(), direction.modZ.toDouble(), dx, dz)

    @JvmStatic
    fun isHeadingTo(cx: Double, cz: Double, dx: Double, dz: Double): Boolean = cx * dx + cz * dz > 0.0

    @JvmStatic
    fun lerp(a: Double, b: Double, t: Double): Double = a + (b - a) * t
}
